### ISSUE TITLE
Window hangs, refuses to close with a `NonSend`/`NonSendMut` system parameter in the `render_app`

### ISSUE THREAD

**Issue by simbleau** [bevyengine/bevy#12912]
## Bevy version

v0.13.2

## \[Optional\] Relevant system information

This took forever to find, but simply having a system that has a
`fn ...(x: NonSend(_)) {}`

added to the render app with default plugins will cause an infinite hang when you try to close the window. Not sure why.

We encountered this in the [`bevy_vello`](https://github.com/loopystudios/bevy_vello) demo, if ran on native platforms.

The line specifically is: 
https://github.com/loopystudios/bevy_vello/blob/dcf1584ead46ad19ff72d7f34110e00f415263f9/src/render/systems.rs#L60

This has been tested and verified on both MacOS and Windows.

**Comment by seabassjh** [comment]
I can reproduce this

**Comment by james7132** [comment]
This makes sense since `NonSend` resources must be dropped on the thread in which they're initialized. With pipelined rendering enabled, the render app/world are sent to a dedicated thread, and ping-pongs between the main thread and the render thread. If you initialized the NonSend resource on the main thread, it shouldn't be accessible on any other thread.

If you can avoid using a `!Send` type as a resource, I strongly suggest avoiding it if possible. If that isn't possible, you may want add it to the main app and use a Send proxy that is used in the render app. This is what `bevy_winit` does.

Long term, we are planning on moving `!Send` resources from the World entirely, moving them into `App` instead. See #9122.

**Comment by simbleau** [comment]
> This makes sense since `NonSend` resources must be dropped on the thread in which they're initialized. With pipelined rendering enabled, the render app/world are sent to a dedicated thread, and ping-pongs between the main thread and the render thread. If you initialized the NonSend resource on the main thread, it shouldn't be accessible on any other thread.
> 
> If you can avoid using a `!Send` type as a resource, I strongly suggest avoiding it if possible. If that isn't possible, you may want add it to the main app and use a Send proxy that is used in the render app. This is what `bevy_winit` does.
> 
> Long term, we are planning on moving `!Send` resources from the World entirely, moving them into `App` instead. See #9122.


> To add to that, subapps are likely to be replaced with a world-based abstraction, rather than a full app, eliminating this footgun completely.

Depending on how long the above solutions will take, this should be documented in `NonSend`/`NonSendMut`. There's no mention of this footgun. ;)

**Comment by alice-i-cecile** [comment]
Yep, totally happy to see a note added to the documentation.

**Comment by rectalogic** [comment]
Internally Bevy uses `NonSend` in `RenderApp` in the `Render` schedule on macos/ios:

https://github.com/bevyengine/bevy/blob/383b3510455c431f34cf3f2c6e3c2d40eddce744/crates/bevy_render/src/view/window/mod.rs#L307-L309

**Comment by msvbg** [comment]
> Internally Bevy uses `NonSend` in `RenderApp` in the `Render` schedule on macos/ios:
> 
> [bevy/crates/bevy_render/src/view/window/mod.rs](https://github.com/bevyengine/bevy/blob/383b3510455c431f34cf3f2c6e3c2d40eddce744/crates/bevy_render/src/view/window/mod.rs#L307-L309)
> 
> Lines 307 to 309 in [383b351](/bevyengine/bevy/commit/383b3510455c431f34cf3f2c6e3c2d40eddce744)
> 
>  #[cfg(any(target_os = "macos", target_os = "ios"))] _marker: Option< 
>      NonSend<bevy_app::NonSendMarker>, 
>  >,

Indeed, ran into this just now. My game hangs on `AppExit` on macOS.

**Comment by msvbg** [comment]
`create_surfaces` appears to be the only possible culprit for this in engine code. `create_surfaces` is added to the `RenderApp` but has a `NonSendMarker` (on macOS/iOS) forcing it to run on the main thread instead of on the render thread. 

So _if_ `create_surfaces` were to get scheduled on shutdown, then the main thread runs `RenderAppChannels::drop`: https://github.com/bevyengine/bevy/blob/8930a157fd7c92d01119c86386ebf7e950e2e76a/crates/bevy_render/src/pipelined_rendering.rs#L59-L63 The main thread is then blocked waiting for the render thread without running main thread tasks. And the render thread is waiting for `create_surfaces`, which needs to run on the main thread, blocking the render thread too.

However... `create_surfaces` has a run condition, so I think the system shouldn't be scheduled on `AppExit`. Which would explain why most of the time, this is a non-issue. But maybe there's an edge case which could cause surface reconfiguration on app exit, thus creating the deadlock. This is about as far as I've come.

**Comment by msvbg** [comment]
Oh hold on, was this maybe fixed in #23838?

EDIT: that PR looks like a more plausible mechanism for macOS freezes than the `create_surfaces` stuff. The original issue still applies for userland code though.
